torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -467,11 +473,8 @@ void ProcessGroupAgent::listenLoop() {
 }
 
 void ProcessGroupAgent::pollTimedOutRPCs() {
-  while (true) {
+  while (rpcRunning_.load()) {
```

Comment:
Don't we need to hold the lock here while checking `rpcRunning`? That's why there was the `while(true)` before

Reply:
@rohan-varma 

The lock was to avoid `::shutdown()` noitifying when the watch thread is not waiting on the cond var.

With the predicate, we can allow the watch dog to miss the notification.
Even if the watch dog thread misses the notification, the predicate will evaluate to `false` (since `rpcRunning_` has already been set to False), the thread will not wait on the cond var, and gets back to the loop condition and decide it deos not need to loop again.